ipfs-8-to-9/migration/swapper.go

```diff
@@ -0,0 +1,165 @@
+type CidSwapper struct {
+	Store ds.Batching // the datastore to migrate.
+}
+
+// Run lists all the keys in the datastore and triggers a swap operation for
+// those corresponding to CIDv1s (replacing them by their raw multihash).
+//
+// Run returns the total number of keys swapped.
+func (cswap *CidSwapper) Run() (uint64, error) {
+	// Always perform a final sync
+	defer cswap.Store.Sync(ds.NewKey("/"))
+	// Query all keys. We will loop all keys
+	// and swap those that can be parsed as CIDv1.
+	queryAll := query.Query{
+		KeysOnly: true,
+	}
+
+	results, err := cswap.Store.Query(queryAll)
+	if err != nil {
+		return 0, err
+	}
+	defer results.Close()
+	resultsCh := results.Next()
+
+	var total uint64
+	var nErrors uint64
+	var wg sync.WaitGroup
+	wg.Add(NWorkers)
+	for i := 0; i < NWorkers; i++ {
+		go func() {
+			defer wg.Done()
+			n, e := cswap.swapWorker(resultsCh)
+			atomic.AddUint64(&total, n)
+			atomic.AddUint64(&nErrors, e)
+		}()
+	}
+	wg.Wait()
+	if nErrors > 0 {
+		return total, errors.New("errors happened during the migration. Consider running it again")
+	}
+
+	return total, nil
+}
+
+// swapWorkers reads query results from a channel and renames CIDv1 keys to
+// raw multihashes by reading the blocks and storing them with the new
+// key. Returns the number of keys swapped and the number of errors.
+func (cswap *CidSwapper) swapWorker(resultsCh <-chan query.Result) (uint64, uint64) {
+	var swapped uint64
+	var errored uint64
+	var curSyncSize uint64
+
+	// Process keys from the results channel
+	for res := range resultsCh {
+		if res.Error != nil {
+			log.Error(res.Error)
+			errored++
+			continue
+		}
+
+		oldKey := ds.NewKey(res.Key)
+		c, err := dsKeyToCid(oldKey)
+		if err != nil {
+			// complain if we find anything that is not a CID but
+			// leave it as it is.
+			log.Log("could not parse %s as a Cid", oldKey)
+			continue
+		}
+		if c.Version() == 0 { // CidV0 are multihashes, leave them.
+			continue
+		}
+
+		// Cid Version > 0
+		newKey := multihashToDsKey(c.Hash())
+		size, err := cswap.swap(oldKey, newKey)
+		if err != nil {
+			log.Error("swapping %s for %s: %s", oldKey, newKey, err)
+			errored++
+			continue
+		}
+		swapped++
+		curSyncSize += size
+
+		// Commit and Sync if we reached SyncSize
+		if curSyncSize >= SyncSize {
+			curSyncSize = 0
+			err = cswap.Store.Sync(ds.NewKey("/"))
+			if err != nil {
+				log.Error(err)
+				errored++
+				continue
+			}
+		}
+	}
+	return swapped, errored
+}
+
+// swap swaps the old for the new key in a single transaction.
+func (cswap *CidSwapper) swap(old, new ds.Key) (uint64, error) {
+	// Unfortunately grouping multiple swaps in larger batches usually
+	// results in "too many open files" errors in flatfs (many really
+	// small files) . So we do small batches instead and Sync at larger
+	// intervals.
+	// Note flatfs will not clear up the batch after committing, so
+	// the object cannot be-reused.
+	batcher, err := cswap.Store.Batch()
```

Comment:
Hm. We could try to detect flatfs. Not sure if there's an easy way to do this now. We might need to extend the datastore interface.

Reply:
If the user has sync enabled, then it gets a nice sync migration. If it is disabled, then both flatfs and badger should be faster. I am not sure batching gives us much on async scenarios but I might be wrong?

Reply:
We don't actually support async updates in flatfs. Well, you _can_ disable "sync", but `Sync` won't do anything. To implement it correctly, we'd need to keep some list of unsynced files.

> I am not sure batching gives us much on async scenarios but I might be wrong?

Batching will likely help with flatfs regardless because it does all writes (open, write, close, etc.) in parallel. But we may be over-optimizing.

Reply:
Let's leave it "unoptimized" for now. I will add some verbosity so that the user sees progress. We can come back to this if it's really really bad.